internal/modules/modules.go

```diff
@@ -139,154 +339,33 @@ func GetModulesForService(ctx context.Context, opts *ModuleResolveOptions) ([]*M
-			if err != nil {
-				return nil, err
-			}
-		} else {
-			// for local + in-memory modules we don't have a version so just stub it
-			version = &resolver.Version{
-				Mutable: true,
-			}
-
-			// if uri is local, represent it as "local" instead of the full path
-			if uriIsLocal(uri) {
-				version.Tag = "local"
-			} else {
-				// otherwise assume in-memory
-				version.Tag = "in-memory"
-			}
-			version.Branch = version.Tag
-
-			// don't attempt to resolve this module again
-			rm.dontResolve = true
-		}
-
-		// if we have a replacement for the module in-memory, use that instead
-		// of creating a new module that's to be resolved
-		if _, ok := opts.Replacements[importPath]; ok {
-			m = opts.Replacements[importPath]
-		} else {
-			var err error
-			m, err = New(ctx, uri, &configuration.TemplateRepository{
-				Name:    importPath,
-				Channel: resolv.conf.Channel,
-				Version: version.GitRef(),
-			})
-			if err != nil {
-				return nil, err
-			}
-		}
-
-		mf, err := m.Manifest(ctx)
-		if err != nil {
-			return nil, err
-		}
-
-		// add the dependencies of this module to the stack to be resolved
-		for i := range mf.Modules {
-			modulesToResolve = append(modulesToResolve, resolveModule{
-				conf:   mf.Modules[i],
-				parent: importPath + "@" + version.String(),
-			})
-		}
-
-		// set the module on our resolved module
-		rm.Module = m
-		rm.version = version
-
-		log.WithFields(logrus.Fields{
-			"module":  importPath,
-			"version": version.GitRef(),
-		}).Debug("resolved module")
-
-		// resolve the next module
-		modulesToResolve = modulesToResolve[1:]
+	return workList{
+		tasks:        modulesToResolve,
+		replacements: strReplacements,
+		log:          opts.Log,
+		resolved:     make(map[string]*resolvedModule),
 	}
-
-	// convert the resolved modules to a list of modules
-	modules := make([]*Module, 0, len(resolved))
-	for _, m := range resolved {
-		modules = append(modules, m.Module)
-	}
-	return modules, nil
 }
 
 // getLatestModuleForConstraints returns the latest module that satisfies the provided constraints
-func getLatestModuleForConstraints(ctx context.Context, uri string, token cfg.SecretData,
-	m *resolveModule, resolved map[string]*resolvedModule) (*resolver.Version, error) {
+func (list *workList) getLatestModuleForConstraints(ctx context.Context, item *workItem, token cfg.SecretData) (*resolver.Version, error) {
 	constraints := make([]string, 0)
-	for _, r := range resolved[m.conf.Name].history {
+	m := item.spec
+
+	list.mu.Lock()
+	// todo(AWinterman): Did i fuckover some modify by reference by lifting this up
+	module, ok := list.resolved[m.conf.Name]
+	// todo(AWinterman): this looks the same as item.inProgressResolution, is it?
+	list.mu.Unlock()
+
+	for _, r := range module.history {
 		if r.constraint != "" {
 			constraints = append(constraints, r.constraint)
 		}
 	}
 
 	channel := m.conf.Channel
-	for _, r := range resolved[m.conf.Name].history {
+	for _, r := range module.history {
```

Comment:
These two calls could panic since you're not checking `ok` until line 386

Reply:
that is true, but as you can sort of see from the dif i just copied this from elsewhere.

Reply:
I think it's actually guaranteed to be in the map by externalities i don't quite understand... 

I can fix if you'd like to eliminate that possibility. Lemme know.

Reply:
okay, fixed.